// jsforce/browser/api/apex.js
import _typeof from "@babel/runtime-corejs3/helpers/typeof";
import _classCallCheck from "@babel/runtime-corejs3/helpers/classCallCheck";
import _createClass from "@babel/runtime-corejs3/helpers/createClass";
import _defineProperty from "@babel/runtime-corejs3/helpers/defineProperty";
import _JSON$stringify from "@babel/runtime-corejs3/core-js-stable/json/stringify";
import "core-js/modules/es.regexp.exec.js";
import "core-js/modules/es.regexp.test.js";
/**
 * @file Manages Salesforce Apex REST endpoint calls
 * @author Shinichi Tomita <shinichi.tomita@gmail.com>
 */
import { registerModule } from '../jsforce';
/**
 * API class for Apex REST endpoint call
 */
export var Apex = /*#__PURE__*/function () {
  /**
   *
   */
  function Apex(conn) {
    _classCallCheck(this, Apex);
    /**
     * Synonym of Apex#delete()
     */
    _defineProperty(this, "del", this.delete);
    this._conn = conn;
  }

  /* @private */
  return _createClass(Apex, [{
    key: "_baseUrl",
    value: function _baseUrl() {
      return "".concat(this._conn.instanceUrl, "/services/apexrest");
    }

    /**
     * @private
     */
  }, {
    key: "_createRequestParams",
    value: function _createRequestParams(method, path, body) {
      var options = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};
      var headers = _typeof(options.headers) === 'object' ? options.headers : {};
      if (!/^(GET|DELETE)$/i.test(method)) {
        headers['content-type'] = 'application/json';
      }
      var params = {
        method: method,
        url: this._baseUrl() + path,
        headers: headers
      };
      if (body) {
        params.body = _JSON$stringify(body);
      }
      return params;
    }

    /**
     * Call Apex REST service in GET request
     */
  }, {
    key: "get",
    value: function get(path, options) {
      return this._conn.request(this._createRequestParams('GET', path, undefined, options));
    }

    /**
     * Call Apex REST service in POST request
     */
  }, {
    key: "post",
    value: function post(path, body, options) {
      var params = this._createRequestParams('POST', path, body, options);
      return this._conn.request(params);
    }

    /**
     * Call Apex REST service in PUT request
     */
  }, {
    key: "put",
    value: function put(path, body, options) {
      var params = this._createRequestParams('PUT', path, body, options);
      return this._conn.request(params);
    }

    /**
     * Call Apex REST service in PATCH request
     */
  }, {
    key: "patch",
    value: function patch(path, body, options) {
      var params = this._createRequestParams('PATCH', path, body, options);
      return this._conn.request(params);
    }

    /**
     * Call Apex REST service in DELETE request
     */
  }, {
    key: "delete",
    value: function _delete(path, options) {
      return this._conn.request(this._createRequestParams('DELETE', path, undefined, options));
    }
  }]);
}();

/*--------------------------------------------*/
/*
 * Register hook in connection instantiation for dynamically adding this API module features
 */
registerModule('apex', function (conn) {
  return new Apex(conn);
});
export default Apex;
